### PR TITLE
dsl generate zip file

### PR DESCRIPTION
**Pull request by gaoning777**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/855)

### CODE REVIEW

**Comment by gaoning777** [comment]
Hold until the backend supports the zip interface.

**Comment by IronPan** [comment]
Tests should pass after https://github.com/kubeflow/pipelines/pull/874 is in

**Comment by Ark-kun** [comment]
Would you mind to call the compiled pipelines 'something.pipeline.zip'?
Basically `s/.zip/.pipeline.zip/g`

**Comment by Ark-kun** [comment]
Also, check the auto-loaded sample list in the Backend. Does it have .tar there?

**Comment by Ark-kun** [comment]
[](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/855/kubeflow-pipeline-e2e-test/1020)	`/test kubeflow-pipeline-e2e-test`

**Comment by Ark-kun** [comment]
Looks like my suspicion was correct: https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/config/sample_config.json#L5

**Comment by IronPan** [comment]
FYI Here is the code that compiles the samples
https://github.com/kubeflow/pipelines/blob/master/backend/Dockerfile#L37

**Comment by Ark-kun** [comment]
I've discovered the complications that were not foreseen: This change will break all existing python pipelines (.py pipelines, notebooks etc).
See https://github.com/kubeflow/pipelines/pull/919 for a fix.

**Comment by hongye-sun** [comment]
Since the backend is in, shall we continue this PR to generate package by file extension?

**Comment by Ark-kun** [comment]
> Since the backend is in, shall we continue this PR to generate package by file extension?

Yes.
AFAIK, the solution we agreed on was:
When the user calls `compile` with a path, that's the exact file path where the compiled pipeline will be stored.
The file format is decided based on the file extension:
'.tar.gz', '.tgz' -> 'Gzipped TAR'
`.zip` -> `ZIP`
? `.yaml`, `.yml` -> `YAML`

File: sdk/python/tests/compiler/compiler_tests.py [hunk 1]
```diff
@@ -118,9 +118,10 @@ def test_operator_to_template(self):
     self.maxDiff = None
     self.assertEqual(golden_output, compiler.Compiler()._op_to_template(op))
 
-  def _get_yaml_from_tar(self, tar_file):
-    with tarfile.open(tar_file, 'r:gz') as tar:
-      return yaml.load(tar.extractfile(tar.getmembers()[0]))
+  def _get_yaml_from_zip(self, zip_file):
```
Comment (hongye-sun) [inline sdk/python/tests/compiler/compiler_tests.py]: Can you add basic tests for covering yaml and tar.gz formats here?

Comment (gaoning777) [reply sdk/python/tests/compiler/compiler_tests.py]: Done.

**Comment by Ark-kun** [comment]
> Could you add a update all readme and notebooks in the whole repo? GCP component samples and readme are under components folder.

Files can be found using `grep -R ".tar.gz'"`

**Pull request merged**